src/Core/Algorithms/Visualization/OsprayDataAlgorithm.cc

```diff
@@ -310,11 +311,156 @@ std::vector<int32_t> OsprayDataAlgorithm::sort_points(EdgeVector edges, std::vec
+  
+  fieldData.dim_x = dimensions_[0];
+  fieldData.dim_y = dimensions_[1];
+  fieldData.dim_z = dimensions_[2];
+  fieldData.origin_x = center.x() - size.x()/2.0;
+  fieldData.origin_y = center.y() - size.y()/2.0;
+  fieldData.origin_z = center.z() - size.z()/2.0;
+  fieldData.spacing_x = size.x()/dimensions_[0];
+  fieldData.spacing_y = size.y()/dimensions_[1];
+  fieldData.spacing_z = size.z()/dimensions_[2];
+  
+  double value;
+  //std::cout << "mname:" << field->mesh()->type_name << std::endl;
+  
+  for (const auto& node : facade->nodes())
+  {
+    auto point = node.point();
+    if (vfield->num_values() > 0)
+    {
+      vfield->get_value(value, node.index());
+      voxels.push_back(value);
+    }
+    vertex_new.push_back(static_cast<float>(point.x()));
+    vertex_new.push_back(static_cast<float>(point.y()));
+    vertex_new.push_back(static_cast<float>(point.z()));
+    
+  }
+  //auto alpha = static_cast<float>(get(Parameters::DefaultColorA).toDouble());
+  if (colorMap)
+  {
+    ColorMap_OSP_helper cmp(colorMap->getColorMapName());
+    obj->tfn.colors = cmp.colorList;
+    
+    // set default opacity for now
+    // obj->tfn.opacities = cmp.opacityList;
+    auto alpha = static_cast<float>(get(Parameters::DefaultColorA).toDouble());
+    obj->tfn.opacities.push_back(alpha);
+    obj->tfn.opacities.push_back(alpha);
+  }
+  fieldData.color = voxels;
+  fieldData.vertex = vertex_new;
+  return obj;
+}
+
+OsprayGeometryObjectHandle OsprayDataAlgorithm::addUnstructVol(FieldHandle field, ColorMapHandle colorMap) const
+{
+  auto obj = makeObject(field);
+  obj->isVolume = true;
+  obj->GeomType="unstructVol";
+  
+  auto& fieldData = obj->data;
+  
+  std::vector<float> voxels;
+  std::vector<float> vertex_new;
+  std::vector<int32_t> index_new;
+  
+  auto facade(field->mesh()->getFacade());
+  auto vfield = field->vfield();
+  auto vmesh = field->vmesh();
+  
+  
+  double value;
+  //std::cout << "mname:" << field->mesh()->type_name << std::endl;
+  for (const auto& node : facade->nodes())
+  {
+    auto point = node.point();
+    if (vfield->num_values() > 0)
+    {
+      vfield->get_value(value, node.index());
+      voxels.push_back(value);
+    }
+    vertex_new.push_back(static_cast<float>(point.x()));
+    vertex_new.push_back(static_cast<float>(point.y()));
+    vertex_new.push_back(static_cast<float>(point.z()));
+  }
+  
+  VMesh::Cell::iterator meshCellIter;
+  VMesh::Cell::iterator meshCellEnd;
+  vmesh->end(meshCellEnd);
+  
+  int numVPerCell = -1;
+  FieldInformation info(field);
+  
+  if(info.is_tetvol()) numVPerCell =4;
+  else if(info.is_hexvol()) numVPerCell =8;
+  else THROW_ALGORITHM_INPUT_ERROR("hex or tet only for unstructured volume!");
+  
+  for (vmesh->begin(meshCellIter); meshCellIter != meshCellEnd; ++meshCellIter)
+  {
+    VMesh::Cell::index_type elemID = *meshCellIter;
+    VMesh::Node::array_type nodesFromCell(8);
```

Comment:
This code can be a bit better, to avoid allocating double the memory for the 4 case versus the 8 case.